From 72b2c32753f161cf6d67b57fe01318774a915d47 Mon Sep 17 00:00:00 2001
From: Kevin McCarthy <kevin@8t8.us>
Date: Mon, 11 Mar 2019 17:38:18 +0800
Subject: [PATCH] Add mutt_buffer_strcpy_n().

---
 buffer.c | 6 ++++++
 buffer.h | 1 +
 2 files changed, 7 insertions(+)

diff --git a/buffer.c b/buffer.c
index e3a306f9..c527de92 100644
--- a/buffer.c
+++ b/buffer.c
@@ -202,6 +202,12 @@ void mutt_buffer_strcpy (BUFFER *buf, const char *s)
   mutt_buffer_addstr (buf, s);
 }
 
+void mutt_buffer_strcpy_n (BUFFER *buf, const char *s, size_t len)
+{
+  mutt_buffer_clear (buf);
+  mutt_buffer_addstr_n (buf, s, len);
+}
+
 
 static void increase_buffer_pool (void)
 {
diff --git a/buffer.h b/buffer.h
index e2184292..72f7fdca 100644
--- a/buffer.h
+++ b/buffer.h
@@ -45,6 +45,7 @@ void mutt_buffer_fix_dptr (BUFFER *);
 /* These two replace the buffer contents. */
 int mutt_buffer_printf (BUFFER*, const char*, ...);
 void mutt_buffer_strcpy (BUFFER *, const char *);
+void mutt_buffer_strcpy_n (BUFFER *, const char *, size_t);
 
 /* These append to the buffer. */
 int mutt_buffer_add_printf (BUFFER*, const char*, ...);
-- 
2.20.1

